[팀 서비스 클론 코딩] 수아(최수현) 미션 제출합니다.

chsua opened the pull request:
# 🎨 1단계 - 팀 서비스의 일부 페이지 클론 코딩

## 🚀 구현한 페이지의 주소와 방식
- 주소: [주소주소](https://chsua-votogether-next-app.vercel.app/PostDetail)
- 렌더링 방식: SSR 

## ✅ 체크리스트

### 선택한 렌더링 방식의 이유
- 아직 next.js에 익숙하지 않아서 next.js에서 기본적으로 추구하는 SSR을 선택했습니다. 하지만 지금 상태에서는 크게 렌더링을 신경쓰고 진행했다고 느껴지지는 않습니다🥲 특히 기존 코드를 옮겨오는 과정이 길었고, 오류가 계속 발생해서 문제해결을 하다보니 next.js도 렌더링도 많이 공부하지 못했습니다ㅠ 

### 코드 재사용
- 기존에 사용하던 컴포넌트는 그대로 가지고 왔습니다. (혹시 그래서 지우긴 했지만 사용하지 않는데도 존재하는 코드가 남아있을 수 있습니다.) 스타일드 컴포넌트와 리액트쿼리를 사용했는데, 그 중 스타일드 컴포넌트는 유지하고, 리액트 쿼리는 기본 fetch로 수정했습니다. 스타일드 컴포넌트가 일부 적용되지 않고 있습니다. 옮긴 컴포넌트가 많기 때문에 스타일드 컴포넌트를 사용하지 않을 수 없었습니다. 리액트쿼리는 사용을 했을 때 정상적으로 불러와지지 않았고, [next에서 리액트쿼리를 사용하지 않아도 된다는 이야기를 봐서 ](https://xionwcfm.tistory.com/339#%F0%9F%98%90%EA%B2%B0%EB%A1%A0-1)과감히 버렸습니다.  

### 안내사항
- 현재 다른 여러 인터렉션은 전부 사용할 수 없도록 막아놓았습니다.
- 데이터를 패치해오는 컴포넌트에서는 스타일드 컴포넌트가 적용되지 않고 있습니다. (로컬에서는 되었는데..🥲)
- 사이트에 장시간 접속해있다면 오류로 인해 에러페이지가 등장하고 있습니다


### 궁금한 점
- next.js나 렌더링에 대한 방식보다는 같은 내용보단 미션을 진행한 크루에게 궁금한 점이 많습니다. 고니는 이번 미션을 하면서 성능이나 제대로 구현(= ssr이나 isr이 이루어지고 있다)되고 있는지 어떤 방식으로 확인하셨나요? 더불어서 서비스마다 다를 것 같은데, 혹시 리액트쿼리나 스타일드 컴포넌트 같은 것들을 사용하셨다면 해당 라이브러리를 그대로 사용하셨는지, 사용하지 않으셨다면 어떻게 하셨는지 궁금합니다.

### (+추가) 갑자기 도메인을 바꾸니까 스타일드 컴포넌트가 전부 적용되네요??? 뭐 fetch가 스타일드 컴포넌트에 영향을 주었나봐요!


# 🧐 2단계 - 렌더링 방식의 분석

## ✅ 체크리스트

### 렌더링 시점 비교
- [ ] SPA와 본인이 선택한 렌더링 방식(SSR, SSG, ISR) 중 본인이 선택한 방식간의 렌더링 시점에 어떤 일이 벌어지는지 설명 (그림 또는 다이어그램 활용도 좋습니다)

### 사용자 경험(UX) 측면 분석
- [ ] 각 렌더링 방식이 사용자 경험에 어떻게 영향을 미치는지 설명

### 성능 측정 및 비교
- [ ] SPA와 본인이 선택한 렌더링 방식(SSR, SSG, ISR)의 성능을 측정하고 비교 (사용한 측정 도구나 방법을 명시)


# 📝 리뷰어를 위한 가이드라인
이번 미션의 핵심 경험은 'Next.js와 프론트엔드에서의 렌더링'을 이해하고 자신의 프로젝트에 적용해보는 것입니다. 이 목표를와 함께 아래 가이드라인을 참고 후 개선을 위한 피드백을 남겨주세요

### 사용자로서, 페이지의 기능을 잘 사용할 수 있는지 학인해주세요,
- [ ] **사용성**: 클론 코딩한 페이지가 원본 페이지와 동일하게 작동하는 지 확인합니다. 의도한 페이지의 기능이 정상적으로 동작하며, 사용자 인터랙션에 따른 반응(예: hover 상태 등)도 잘 구현되었는 지 점검합니다. 이를 통해 사용자에게 어떤 경험이 더 최적화 될 수 있는지를 피드백해주세요.

### Next.js에서 제공하는 기능을 적극적으로 활용했는지 확인해 주세요 
> 아래 가이드라인 외에도 Next.js가 제공하는 기능에 대해 피드백 남겨주세요.
- [ ] **데이터 요청**: getServerSideProps나 getStaticProps 같은 데이터 fetching 함수를 올바르게 사용했는지 확인합니다.
- [ ] **파일 기반 라우팅**: Next.js의 파일 기반 라우팅 시스템을 잘 활용했는지 점검합니다.
- [ ] **에러 핸들링**: 네트워크 요청 실패나 예외 처리 등 에러 상황에서 앱이 적절히 반응하는지 파악합니다. 예를 들어 잘못된 URL 경로로 접근 시 404 페이지를 보여주는 등의 에러 핸들링을 확인해봅니다.

chsua committed:
chore: 최초 next.js 설치

chsua committed:
chore: 파일 정리 및 예제 페이지 작성

chsua committed:
feat: 게시글 상세 페이지 관련 파일 이동

chsua committed:
feat: mock 데이터 생성

chsua committed:
fix: function 함수정의 -> 화살표 함수로 수정

- json에서 에러가 발생, next.js오류로 화살표 함수로 수정하면 된다고 하여 수정

chsua committed:
refactor: 사용하지 않는 코드 삭제

chsua committed:
feat: 카테고리 목록 mock data 추가 및 댓글 목록 데이터 수정

chsua committed:
feat: 글로벌 스타일 추가 및 "use client" 코드 추가

chsua committed:
feat: 사용하지 않는 코드 및 파일 삭제

chsua committed:
chore: 환경설정

chsua committed:
feat: 카테고리 목록 fetch => mock data로 교체

chsua committed:
feat: 글로벌 스타일 설정

chsua committed:
feat: alert 함수 안에 인수 넣어주기

chsua committed:
feat: 게시글 상세 페이지에서 레이아웃 적용

chsua committed:
fix: 스타일드 컴포넌트가 적용되지 않는 오류 수정

chsua committed:
feat: 적절한 태그로 수정

chsua committed:
feat: 리액트쿼리 대신 컴포넌트에서 직접 fetch하도록 수정

chsua committed:
refactor: 불필요한 파일 삭제 및 util/type 코드 분리

chsua committed:
refactor: 불필요한 파일 삭제

chsua committed:
chore: 배포 스크립트 추가

chsua committed:
chore: 배포 스크립트 삭제

jeongwusi commented on app/page.tsx:
```diff
@@ -0,0 +1,5 @@
+ const Home =()=> {
+  return <main>homepage</main>;
+}
+
+export default Home
```
![image](https://github.com/woowacourse/frontend-rendering/assets/122500517/1674988f-5c41-4828-837c-30dd6d343748)

SSR을 이용하려면 page.tsx를 메인 페이지로 구성을 해야 하는 거 같더라고요!! 저 같은 경우는 page.tsx를 중심으로 서버 컴포넌트와 클라이언트 컴포넌트로 나눠서 조합을 했습니당

chsua replied:
제가 상세페이지는 실제 페이지에서 메인(홈)페이지가 아니라서 home은 예시로만 작성하고 비워두었습니다!
혹시 요 코멘트가 어떤걸 변경하라고 말씀주신걸까요??😮

jeongwusi replied:
음... 이번 미션에서는 상세 페이지가 결국 메인 페이지이기 때문입니다. 이번 미션에서 페이지가 결국 1개 아닌가욤?? 😅

그렇기 때문에 `page.tsx`에서 코드를 작성하는 것이 구조적 완성도나 코드 안정성 측면에서 도움이 될 거라고 생각하여 남긴 코멘트였습니다. ^^

jeongwusi commented on app/PostDetail/PostDetail/index.tsx:
```diff
@@ -0,0 +1,130 @@
+  );
+};
+
+export default PostDetail;
```
지금 이 부분에서 메인 페이지 역할을 하고 있는데 SSR을 이용하려면 이 부분을 app/page.tsx에서 해야 합니당

chsua replied:
근데 그러면 ssr로 렌더링하는 경우 홈페이지로 진입하지 않고 다른 루트로 들어오는 경우엔 ssr이 적용되지 않는건가요?? 
아니면 홈페이지에 연결만 되어있으면 홈페이지로 진입하지 않고 다른 루트로 들어오는 경우에도 ssr이 적용되는 건가요?

jeongwusi replied:
수아가 제출해주신 미션도 SSR로 이루어졌기 때문에 서버 컴포넌트 내에서 클라이언트 컴포넌트를 구성을 해야된다고 생각을 했습니다. 

지금 메인페이지로 사용되고 있는 `app/PostDetail/PostDetail/index.tsx` 이 부분이 `page.tsx`로 구성이 안되어 있기 때문에 서버 컴포넌트로써 제대로 역할을 못할 수 있다고 생각하였습니다.

때문에 메인 페이지로 사용되고 있는 상세페이지를 `app/page.tsx`에서 코드를 작성 해야한다는 의미였습니당

jeongwusi commented on app/PostDetail/PostDetail/index.tsx:
```diff
@@ -0,0 +1,130 @@
+  transformPostResponse,
+} from "@/utils/transType";
+
+const BASE_URL = process.env.NEXT_PUBLIC_ENV_VOTOGETHER_BASE_URL;
```
지금도 배포 페이지의 통신이 잘 안되는 거 같은데... Vercel에 환경 변수 등록 해주셨나욤??

chsua replied:
아ㅠㅠㅠ 죄송합니다ㅠㅠ 지금 백엔드가 서버에서 무중단배포 등을 적용하는 중이라 서버가 다운되었다가 복구되었다가 반복되고 있습니다ㅠㅠ

jeongwusi replied:
오옹오옹 넵넵 알겠사욥니당

jeongwusi commented on app/PostDetail/PostDetail/index.tsx:
그리고 Next.js 13의 App Router에서 서버 컴포넌트를 사용하실 경우 `index.tsx` 등등 사용할 수 없는 것으로 알고 있어용

page.tsx를 사용을 해야 하는 거 같아여

<img width="1258" alt="스크린샷 2023-10-13 오전 11 56 56" src="https://github.com/woowacourse/frontend-rendering/assets/122500517/47ccefc8-95b5-4de3-b1fb-ada438169ea3">

chsua replied:
app 바로 하위 페이지말고도 모든 컴포넌트를 다 페이지로 이름을 바꿔야 하나요..?
페이지 안에 포함되어있는 컴포넌트라고 생각이 들어 ```index.tsx```로 처리했는데 모두 page로 사용해야 하는 건가요??

```
// 현재 app 폴더
app
 ┣ PostDetail
 ┃ ┣ BottomButtonPart
 ┃ ┃ ┣ index.tsx : 페이지 내 컴포넌트
 ┃ ┃ ┗ style.ts
 ┃ ┣ InnerHeaderPart
 ┃ ┃ ┣ index.tsx : 페이지 내 컴포넌트
 ┃ ┃ ┗ style.ts
 ┃ ┣ PostDetail : 페이지 내 컴포넌트
 ┃ ┃ ┣ index.tsx
 ┃ ┃ ┗ style.ts
 ┃ ┣ page.tsx  : 현재 페이지
 ┃ ┗ types.ts 
 ┣ layout.tsx
 ┣ page.tsx
 ┗ registry.tsx
```

jeongwusi replied:
`BottomButtonPart`, `InnerHeaderPart`, `PostDetail` 이 부분은 일반적인 클라이언트 사이드 컴포넌트이기에 당연히 `page.tsx`로 변경하실 필요 없습니당ㅎㅎ

`PostDetail/page.tsx`만 page로 구성하시면 될 거 같아용!! 잘 해주신 거 같습니닷 👍🏻👍🏻👍🏻

jeongwusi commented on postcss.config.js:
```diff
@@ -0,0 +1,6 @@
+module.exports = {
+  plugins: {
+    tailwindcss: {},
+    autoprefixer: {},
+  },
+}
```
Styled Components를 사용하시니 삭제를 해도 될 거 같아욤!

jeongwusi commented on app/registry.tsx:
```diff
@@ -0,0 +1,29 @@
+"use client";
+
+import React, { useState } from "react";
+import { useServerInsertedHTML } from "next/navigation";
+import { StyleRegistry, createStyleRegistry } from "styled-jsx";
+import { ThemeProvider } from "styled-components";
+import { theme } from "@/styles/theme";
+import { GlobalStyle } from "@/styles/globalStyle";
+
+const StyledJsxRegistry = ({ children }: { children: React.ReactNode }) => {
+  const [jsxStyleRegistry] = useState(() => createStyleRegistry());
+
+  useServerInsertedHTML(() => {
+    const styles = jsxStyleRegistry.styles();
+    jsxStyleRegistry.flush();
+    return <>{styles}</>;
+  });
+
+  return (
+    <StyleRegistry registry={jsxStyleRegistry}>
+      <ThemeProvider theme={theme}>
+        <GlobalStyle />
+        {children}
+      </ThemeProvider>
+    </StyleRegistry>
+  );
+};
+
+export default StyledJsxRegistry;
```
오홍...! Styled Components 환경 설정인가 보네영 (신기신기)

근데 그 [공식 문서](https://nextjs.org/docs/app/building-your-application/styling/css-in-js#styled-components)에서 되어 있는 거랑은 조금 다르네영?! 같은 방식인가영?

chsua replied:
이상하게 저는  ```if (typeof window !== 'undefined') return <>{children}</>``` 코드가 있으면 에러가 발생했습니다😢😢😢
공식문서에서 말하기를 window가 있으면 해당 오류가 난다고 하여 삭제했습니다.

chsua replied:
오 오류가 나서 수정한 부분이 있는데 다시 공식문서를 보고 설정했더니 됩니다! 처음에 할때 다른 부분이랑 부딪혀서 안됐던 건가봐요!

jeongwusi replied:
됐다고 하니 다행이네욤 😊

jeongwusi commented on utils/fetch.ts:
```diff
@@ -0,0 +1,22 @@
+export const getFetch = async <T>(
```
오류가 나오는게 이 부분이 문제 일수도?!

그 서버 컴포넌트에서는 fetch를 지원해주기는 하지만 지금 실질적으로 fetch를 하는 부분이 아예 app 폴더 안에 없고 utils로 완전히 떨어져 나왔네용!

이 부분을 아예 각각의 컴포넌트에 넣어주는 방식을 채택 또는 저 처럼 Server Actions를 고려를 해봐도 될 거 같아용 ^^

jeongwusi commented on components/SearchBar/style.ts:
```diff
@@ -0,0 +1,54 @@
+"use client";
```
제가 Next.js에서 Styled Components를 사용을 안해봐서 잘 모르겠는데...ㅎㅎ 스타일 관련된 파일에는 use clinet를 사용을 하셨는데, 해당 파일과 관련된 컴포넌트 파일의 최상단에 use client가 없더라구여 

App Router 같은 경우 use client가 없으면 전부 서버 컴포넌트로 인식하는 것으로 알고 있어욤

jeongwusi reviewed (changes requested):
안녕하세요 수아~!!

바쁘신데도 불구하고 미션을 멋지게 해주셨네요 💯

질문주신 내용과 더불어 몇가지 코멘트를 달아서 리뷰드려욤 😁

### 이번 미션을 하면서 성능이나 제대로 구현(= ssr이나 isr이 이루어지고 있다)되고 있는지 어떤 방식으로 확인하셨나요?
<img width="1434" alt="스크린샷 2023-10-13 오전 11 09 53" src="https://github.com/woowacourse/frontend-rendering/assets/122500517/4b3f8850-0b45-4786-a797-5311d869d213">

<img width="905" alt="스크린샷 2023-10-13 오전 11 11 13" src="https://github.com/woowacourse/frontend-rendering/assets/122500517/208e5fc1-9edd-460b-bce5-825e7dd81923">

저 같은 경우는 배포 후 네트워크 탭에서 직관적으로 확인할 수 있었습니다. 배포 후 정적 페이지인 HTML 문서가 먼저 다운로드가 되고 그 이후 이미지가 다운로드가 되더라구여!! (아마... 이미지는 NEXT에서 최적화 작업을 하는 것으로 알고 있어욤 ^^) 

그래서 Lighthouse 점수 보시면 아무런 최적화 작업 한 것이 없는데... 100점이 나옵니다 ㅋㅋㅋ (LCP가 0.3초라능...😇) 왜 주변에서 SSR 노래를 부르는 지 알 수 있었습니다. ㅎㅎ

### 리액트쿼리나 스타일드 컴포넌트 같은 것들을 사용하셨다면 해당 라이브러리를 그대로 사용하셨는지, 사용하지 않으셨다면 어떻게 하셨는지 궁금합니다.

저희 팀 같은 경우 프로젝트에서는 React Query와 Styled Components를 사용해서 작업을 하였습니다. 다만 Next.js의 서버 컴포넌트의 경우 CSS Module과 Tailwind CSS만 지원을 해주는 것으로 알고 있습니다. 물론 CSS-in-JS의 경우 클라이언트 컴포넌트를 이용하여 사용할 수 있지만, 저는 이번 미션에서는 Next.js의 온전하게 사용을 해보기 위해 학습 측면에서 CSS Module로 마이그레이션을 하였습니다.

네트워크 통신의 경우 React Query에서 지원해주는 [Server Rendering & Hydration](https://tanstack.com/query/v5/docs/react/guides/ssr)을 통해 SSR로 통신을 해줄 수 있는 것으로 알고 있습니다. (물론 클라이언트 컴포넌트를 이용해서 일반적으로 저희가 구현한 방식으로도 가능합니다.)

다만 저는 이번에도 학습 측면에서 다른 방법으로 통신을 하고 싶었습니다. 그래서 저는 Next.js 13.4부터 지원을 해준 [Server Actions](https://nextjs.org/docs/app/api-reference/functions/server-actions)를 사용해서 네트워크 통신을 하였습니다.

Server Actions를 통해 fetch 관련된 로직을 분리 할 수 있었으며, 동적으로 받아오는 로직들을 SSR 형식으로 받아올 수 있었습니다. 이와 관련해서는 제가 [블로그 글](https://velog.io/@wusi-hub/Next.js와-Server-Actions를-활용한-웹-애플리케이션의-성능-향상)에 정리하기도 하였습니다. ^_^;

이밖에 저는 이번 미션을 하면서 느낀 점은 어느 부분은 SSR로 할 지 또는 CSR로 할 지에 대한 설계를 잘 해야 될 거 같다라는 느낌을 받았습니다. ㅎㅎ

제가 미션 하면서 참고한 영상이 있는데 수아도 도움이 될 거 같아서 공유드립니다. 

- https://youtu.be/TQQPAU21ZUw?si=LsJ4nYNYaDBpxcDB (Dan Abramov 선생님이 뭐 알려줘여 ㅋㅋ)

그럼... 바쁘실텐데 너무 고생많으셨어용~!~! 간단한 코멘트 확인 부탁드려욤 >_<

감사합니다. 🙇🏻

chsua committed:
chore: 사용하지 않는 플러그인(tailwindcss) 삭제

chsua committed:
chore: svg 이미지 처리방식 설정

chsua committed:
fix: url이 잘못 설정되어있어 발생하는 오류 수정

- 기존 서비스에서는 domain/postId인데, 해당 코드에서는 domain/PostDetail로 설정했었음
- 이로 인해 오류가 발생하여 기존 url과 동일하게 수정

chsua committed:
chore: svg 바벨 플러그인 설치

chsua committed:
fix: 스타일드컴포넌트 적용 공식문서에 따라 수정

chsua committed:
fix: 해당 플러그인을 사용하면 svg가 인식이 되지 않음

chsua commented:
### 몇 가지 수정했습니다!
1. url로 인한 object object 오류 수정
2. 스타일드 컴포넌트 적용

### 다만 아직 문제가 발생합니다ㅠㅠ
1. 처음에 svg를 적용하려고 하려고 플러그인을 설치하였으나 그럼 svg가 인식이 안되서 문제가 발생했습니다. 그래서 일단 삭제했습니다.
2. 이상하게 갑자기 무한 fetch가 일어나는 경우가 발생합니다만 무엇이 문제인지 모르겠습니다. 다른 사람들 사이트도 종종 그런 것 같습니다.

 ### 코멘트 달았습니다
그 데이터가 안오는 오류는 fetch의 문제는 아니고 저희 개발서버를 고치는 중이라 간헐적으로 서버가 내려가고 있습니다. 미리 말씀 못드려서 죄송합니다.
지금 상태에서는 처음에 클라이언트 컴포넌트를 제외하고 html이 css까지 입혀서 오는 것을 확인했습니다!

jeongwusi reviewed (approved):
안녕하세요 수아~!~! 

주말 잘 보내고 계신가욤??

미션 리팩토링 하시느라 고생 많으셨습니당 ^_^

수아가 남겨준 코멘트에 대해서 간단하게 답변 남겼습니다. 확인 해주시면 될 거 같습니닷 ㅎㅎ

그럼... 남은 주말도 행복하게 보내시고 2단계 미션에서 뵙도록 하겠습니다. 😄

감사합니다. 🙇🏻
